feat(portal): allow socket based postgres connections

oddlama opened the pull request:
This allows connections to the postgresql database via the standard socket, which - opposed to TCP sockets - allows `peer` authentication based on local unix users. This removes the need for a password and is much simpler to deploy when running components locally.

In the current form, `DATABASE_SOCKET_DIR` takes precedence over hostname, if the environment variable is present. I found that `compile_config!` somehow enforces a value to be present which is explicitly not what I want for some of these values (i think). I'd be glad if anyone with more elixir experience can guide me as to how I can make this more idiomatic.

oddlama committed:
feat(portal): allow socket based postgres connections

jamilbk commented on elixir/config/runtime.exs:
```diff
@@ -8,15 +8,17 @@ if config_env() == :prod do
+    {:ssl_opts, compile_config!(:database_ssl_opts)},
+    {:parameters, compile_config!(:database_parameters)}
+  ]
+  ++ (if System.get_env("DATABASE_PASSWORD"), do: [{:password, compile_config!(:database_password)}], else: [])
```
Hm, I think we had moved away from using `System.get_env/3` directly in configuration because it creates a difficult to reason about mess.

What do you think of instead, adding a `compile_config/3` function similar to [this one](https://github.com/firezone/firezone/blob/main/elixir/apps/domain/lib/domain/config.ex#L35) but returns `nil` instead for missing keys? Then your conditional here would use the result of that, and not the System ENV directly.

Based on that function's `@doc` comment it appears that the original author planned to write one anyhow.

oddlama replied:
I've added this now, could you check whether my implementation is what you had intended?

jamilbk reviewed (commented):
Suggestion to add a `compile_config/3` function that does not raise, but instead returns `nil`, that you can use in your conditional to specify DB parameters.

jamilbk commented on elixir/apps/domain/lib/domain/config.ex:
```diff
@@ -42,6 +42,24 @@ defmodule Domain.Config do
   end
 
+  @doc """
+  Similar to `compile_config/2` but returns nil if the configuration is invalid.
```
```suggestion
  Similar to `compile_config!/3` but returns nil if the configuration is invalid.
```

jamilbk commented on elixir/apps/domain/lib/domain/config.ex:
```diff
@@ -42,6 +42,24 @@ defmodule Domain.Config do
+      {:ok, _source, value} ->
+        value
+
+      {:error, reason} ->
```
```suggestion
      {:error, _reason} ->
```

jamilbk reviewed (approved):
Just a couple more minor things otherwise LGTM.

oddlama committed:
feat(portal): add configuration for the healthz port (#8020)

This is required to run multiple components on a single machine (even if
the processes are sandboxed), since they will share a network namespace
and thus cannot bind to the same port.

Currently port `4000` is hardcoded, this PR allows this to be configured
by an environment variable.

jamilbk committed:
Apply suggestions from code review

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
Fix formatting

jamilbk committed:
Use default port in compile config

jamilbk committed:
Merge branch 'contrib' into postgres-socket
